File: go/internal/cmd/body.go
package cmd

import (
	"fmt"
	"io"
	"os"

	"github.com/spf13/cobra"

	"git.zib.de/cspiegel/threads/internal/git"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	bodySet    bool
	bodyAppend bool
	bodyCommit bool
	bodyMsg    string
)

var bodyCmd = &cobra.Command{
	Use:   "body <id>",
	Short: "Edit Body section (stdin for content)",
	Long: `Edit the Body section of a thread.

Content is read from stdin. Use --set to replace or --append to add.`,
	Args: cobra.ExactArgs(1),
	RunE: runBody,
}

func init() {
	bodyCmd.Flags().BoolVar(&bodySet, "set", false, "Replace body content")
	bodyCmd.Flags().BoolVar(&bodyAppend, "append", false, "Append to body content")
	bodyCmd.Flags().BoolVar(&bodyCommit, "commit", false, "Commit after editing")
	bodyCmd.Flags().StringVarP(&bodyMsg, "m", "m", "", "Commit message")
}

func runBody(cmd *cobra.Command, args []string) error {
	ws := getWorkspace()
	ref := args[0]

	// Default to set mode
	if !bodySet && !bodyAppend {
		bodySet = true
	}

	// Read content from stdin
	var content string
	stat, _ := os.Stdin.Stat()
	if (stat.Mode() & os.ModeCharDevice) == 0 {
		data, err := io.ReadAll(os.Stdin)
		if err != nil {
			return err
		}
		content = string(data)
	}

	if content == "" {
		return fmt.Errorf("no content provided (use stdin)")
	}

	file, err := workspace.FindByRef(ws, ref)
	if err != nil {
		return err
	}

	t, err := thread.Parse(file)
	if err != nil {
		return err
	}

	if bodySet {
		t.Content = thread.ReplaceSection(t.Content, "Body", content)
	} else {
		t.Content = thread.AppendToSection(t.Content, "Body", content)
	}

	if err := t.Write(); err != nil {
		return err
	}

	mode := "set"
	if bodyAppend {
		mode = "append"
	}
	fmt.Printf("Body %s: %s\n", mode, file)

	if bodyCommit {
		msg := bodyMsg
		if msg == "" {
			msg = git.GenerateCommitMessage(ws, []string{file})
		}
		if err := git.AutoCommit(ws, file, msg); err != nil {
			return err
		}
	} else {
		fmt.Printf("Note: Thread %s has uncommitted changes. Use 'threads commit %s' when ready.\n", ref, ref)
	}

	return nil
}

File: rust/src/cmd/body.rs
use std::io::{self, Read};
use std::path::Path;

use clap::Args;
use clap_complete::engine::ArgValueCompleter;

use crate::git;
use crate::thread::{self, Thread};
use crate::workspace;

#[derive(Args)]
pub struct BodyArgs {
    /// Thread ID or name reference
    #[arg(add = ArgValueCompleter::new(crate::workspace::complete_thread_ids))]
    id: String,

    /// Replace body content
    #[arg(long)]
    set: bool,

    /// Append to body content
    #[arg(long)]
    append: bool,

    /// Commit after editing
    #[arg(long)]
    commit: bool,

    /// Commit message
    #[arg(short = 'm', long)]
    m: Option<String>,
}

pub fn run(args: BodyArgs, ws: &Path) -> Result<(), String> {
    // Default to set mode
    let set_mode = args.set || !args.append;

    // Read content from stdin
    let content = read_stdin_if_available();

    if content.is_empty() {
        return Err("no content provided (use stdin)".to_string());
    }

    let file = workspace::find_by_ref(ws, &args.id)?;

    let mut t = Thread::parse(&file)?;

    if set_mode {
        t.content = thread::replace_section(&t.content, "Body", &content);
    } else {
        t.content = thread::append_to_section(&t.content, "Body", &content);
    }

    t.write()?;

    let mode = if set_mode { "set" } else { "append" };
    println!("Body {}: {}", mode, file.display());

    if args.commit {
        let msg = args.m.unwrap_or_else(|| {
            git::generate_commit_message(ws, &[file.to_string_lossy().to_string()])
        });
        git::auto_commit(ws, &file, &msg)?;
    } else {
        println!(
            "Note: Thread {} has uncommitted changes. Use 'threads commit {}' when ready.",
            args.id, args.id
        );
    }

    Ok(())
}

fn read_stdin_if_available() -> String {
    // Check if stdin has data available
    if is_stdin_piped() {
        let mut buffer = String::new();
        if io::stdin().read_to_string(&mut buffer).is_ok() {
            return buffer;
        }
    }
    String::new()
}

fn is_stdin_piped() -> bool {
    use std::io::IsTerminal;
    !io::stdin().is_terminal()
}
